mgm/src/encdec.rs

```diff
@@ -0,0 +1,305 @@
+use crate::{
+    gf::GfElement,
+    sealed::{Counter, Sealed},
+    DecArgs, EncArgs, MgmBlockSize,
+};
+use aead::{
+    generic_array::{
+        typenum::{Unsigned, U16, U8},
+        GenericArray,
+    },
+    Error,
+};
+use cipher::{Block, BlockEncrypt, ParBlocks};
+use subtle::ConstantTimeEq;
+
+pub(crate) fn encrypt<C, E64, E128>(args: EncArgs<'_, C>) -> Result<Block<C>, Error>
+where
+    C: BlockEncrypt,
+    C::BlockSize: MgmBlockSize,
+    E64: GfElement<N = U8>,
+    E128: GfElement<N = U16>,
+{
+    // ideally we would use type-level branching here
+    match C::BlockSize::USIZE {
+        8 => encrypt_inner::<C, E64>(args),
+        16 => encrypt_inner::<C, E128>(args),
+        _ => unreachable!(),
```

Comment:
This part is somewhat ugly and breaks type-level safety of `encrypt_inner` and `decrypt_inner` functions (i.e. they may panic if incorrect `GfElement` will be provided to them), but AFAIK Rust currently does not have a nice way of doing type-level branching.